go/libraries/doltcore/sqle/rowexec/builder.go

```diff
@@ -0,0 +1,362 @@
+
+import (
+	"context"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/expression"
+	"github.com/dolthub/go-mysql-server/sql/plan"
+
+	"github.com/dolthub/dolt/go/libraries/doltcore/doltdb"
+	"github.com/dolthub/dolt/go/libraries/doltcore/doltdb/durable"
+	"github.com/dolthub/dolt/go/libraries/doltcore/schema"
+	"github.com/dolthub/dolt/go/libraries/doltcore/sqle"
+	"github.com/dolthub/dolt/go/libraries/doltcore/sqle/index"
+	"github.com/dolthub/dolt/go/store/prolly"
+	"github.com/dolthub/dolt/go/store/prolly/tree"
+	"github.com/dolthub/dolt/go/store/val"
+)
+
+type Builder struct{}
+
+var _ sql.NodeExecBuilder = (*Builder)(nil)
+
+func (b Builder) Build(ctx *sql.Context, n sql.Node, r sql.Row) (sql.RowIter, error) {
+	switch n := n.(type) {
+	case *plan.JoinNode:
+		if n.Op.IsLookup() && !n.Op.IsPartial() {
+			if ita, ok := getIta(n.Right()); ok && len(r) == 0 && simpleLookupExpressions(ita.Expressions()) {
+				if _, _, dstIter, _, dstTags, dstFilter, err := getSourceKv(ctx, n.Right(), false); err == nil && dstIter != nil {
+					if srcMap, srcIter, _, srcSchema, srcTags, srcFilter, err := getSourceKv(ctx, n.Left(), true); err == nil && srcSchema != nil {
+						if keyLookupMapper := newLookupKeyMapping(ctx, srcSchema, srcMap, dstIter.InputKeyDesc(), ita.Expressions()); keyLookupMapper.valid() {
+							split := len(srcTags)
+							projections := append(srcTags, dstTags...)
+							rowJoiner := newRowJoiner([]schema.Schema{srcSchema, dstIter.Schema()}, []int{split}, projections, dstIter.NodeStore())
+							return rowIterTableLookupJoin(srcIter, dstIter, keyLookupMapper, rowJoiner, srcFilter, dstFilter, n.Filter, n.Op.IsLeftOuter(), n.Op.IsExcludeNulls())
+						}
+					}
+				}
+			}
+		}
+	default:
+	}
+	return nil, nil
+}
+
+func getIta(n sql.Node) (*plan.IndexedTableAccess, bool) {
+	switch n := n.(type) {
+	case *plan.TableAlias:
+		return getIta(n.Child)
+	case *plan.Filter:
+		return getIta(n.Child)
+	case *plan.IndexedTableAccess:
+		return n, true
+	default:
+		return nil, false
+	}
+}
+
+// simpleLookupExpressions returns true if |keyExprs| includes only field
+// references and literals
+func simpleLookupExpressions(keyExprs []sql.Expression) bool {
+	for _, e := range keyExprs {
+		switch e.(type) {
+		case *expression.Literal, *expression.GetField:
+		default:
+			return false
+		}
+	}
+	return true
+}
+
+// prollyToSqlJoiner converts two KV pairs into a sql row
+type prollyToSqlJoiner struct {
+	ns tree.NodeStore
+	// kvSplits are offsets between consecutive kv pairs
+	kvSplits    []int
+	desc        []kvDesc
+	ordMappings []int
+}
+
+type kvDesc struct {
+	keyDesc     val.TupleDesc
+	valDesc     val.TupleDesc
+	keyMappings []int
+	valMappings []int
+}
+
+func newRowJoiner(schemas []schema.Schema, splits []int, projections []uint64, ns tree.NodeStore) *prollyToSqlJoiner {
+	numPhysicalColumns := getPhysicalColCount(schemas, splits, projections)
+
+	// | k1 | v1 | k2 | v2 | ... | ords |
```

Comment:
What is "ords" here?

Reply:
There's two steps of the mapping, from KV to schema order, and then to output projection row. I'm not sure if we really need two steps, but I just refactored the existing code to generalize to multiple tuples.

Reply:
Alright. This comment is still confusing. It's not really clear to someone who doesn't already understand the structure of the data.